src/OrleansRuntime/Startup/StartupBuilder.cs

```diff
@@ -30,30 +44,32 @@ internal static IServiceProvider ConfigureStartup(string startupTypeName, out bo
                 servicesMethod = FindConfigureServicesDelegate(startupType);
                 if (servicesMethod != null && !servicesMethod.MethodInfo.IsStatic)
                 {
-                    usingCustomServiceProvider = true;
+                    hasValidServiceBuilderMethod = true;
                 }
             }
-
-            RegisterSystemTypes(serviceCollection);
-
-            if (usingCustomServiceProvider)
+            Func<IServiceCollection, IServiceProvider> serviceProviderBuilder = null;
+            if (hasValidServiceBuilderMethod)
             {
                 var instance = Activator.CreateInstance(startupType);
-                return servicesMethod.Build(instance, serviceCollection);
+                serviceProviderBuilder = serviceCollection => servicesMethod.Build(instance, serviceCollection);
             }
 
-            return serviceCollection.BuildServiceProvider();
+            return ConfigureStartup(serviceProviderBuilder, configureServices, out usingCustomServiceProvider);
         }
 
-        private static void RegisterSystemTypes(IServiceCollection serviceCollection)
+        private static IServiceProvider ConfigureStartup(Func<IServiceCollection, IServiceProvider> serviceProviderBuilder, Action<IServiceCollection> configureServices, out bool usingCustomServiceProvider)
         {
-            // Register the system classes and grains in this method.
-            // Note: this method will probably have to be moved out into the Silo class to include internal runtime types.
+            IServiceCollection serviceCollection = new ServiceCollection();
+            configureServices(serviceCollection);
 
-            serviceCollection.AddTransient<GrainBasedMembershipTable>();
-            serviceCollection.AddTransient<GrainBasedReminderTable>();
-        }
+            usingCustomServiceProvider = serviceProviderBuilder != null;
+            if (usingCustomServiceProvider)
+            {
+                return serviceProviderBuilder(serviceCollection);
+            }
```

Comment:
This might be a nice first step for now.
Nevertheless. we should probably start splitting the registrations into helper methods, so that we can eventually do something like in AspNet MVC, where the user decides what from the framework to put into the container. Most of the times that requires the user to add an explicit 1 line config call such as `services.AddMvc()` (which adds every default implementation in MVC) but allows users to pick and choose different implementations of some of the internal services if they want. For more info of what I mean: https://docs.asp.net/en/latest/fundamentals/dependency-injection.html

With the current approach, if Orleans' defaults are always registered, there is not an easy way for opting for a different implementation unless we rely on the nasty approach of specifying the concrete type name that will be resolved, by having explicit configuration knobs (type names) for each internal dependency we might want replaceable.